Add GPL family of Licenses

gabrielgellner opened the pull request:
Pull request to resolve #46

tkelman commented:
We should probably default to the + "or later" versions of these for wider compatibility.

johnmyleswhite commented:
If we do this, we should try to make it clear that many of us *strongly* encourage the use of non-GPL licenses.

gabrielgellner commented:
@tkelman I meant to! Would this mean use a different license text (as these are the only versions I find on the GNU page), and then edit the top output text to use the suggested text from the bottom of the license file (quoted at bottom of this comment)? I was worried about editing the raw license file ;)

@johnmyleswhite I have not skin in the politics of this and am happy to add any suggestions, I just think it is wrong have a basic tool make such a strong proclamation when a lot of code might want/need to be GPL. What do you think the best way to include this warning might be? It does seem that the mailing lists are a pretty harsh crucible for using GPL ;)

> This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

johnmyleswhite commented:
Somebody with less skin in the game than me should take over, but I'd say something to the effect of:

> If you release code for Package X under the GPL, you may discourage collaboration from members of the Julia community who work on non-GPL packages. For example, if a user works on Package Y, which is licensed under the MIT license that is used in many community projects, they can not safely read the code you contribute to Package Y because any indication that their work is derivative could lead to litigation. In effect, you create a situation in which your source code is completely closed to anyone who is working on a non-GPL project.

carlobaldassi commented:
@johnmyleswhite IANAL, but to me that wording seems definitely excessive. GPL is contagious, but not looking at someone's code because they have released some other code under GPL at some point is absurd, in my opinion.

johnmyleswhite commented:
That isn't what I said, Carlos. You can't read their GPL code. Many people do not realize that reading GPL code and then writing non-GPL code is a problem.

gabrielgellner commented:
@tkelman I have fixed the licenses to actually be what they should be. That is I have removed the directions on the use of the license, and added the suggested text (which is version+) to the top so that the generated copyright information will make the included license.md a proper declaration of intent.

@johnmyleswhite @carlobaldassi I have added the suggested paragraph to the `Readme.md`. It might be a bit excessive, but I think it does capture the general feelings of the Julia community, right or wrong. Personally I don't find the note that off putting AND this is a net movement forward. It keeps it clear that BSD-compatible licenses are the *very* strongly preferred, but at least we give the user the choice to use the license they want.

carlobaldassi commented:
>  if a user works on Package Y, which is licensed under the MIT license that is used in many community projects, they can not safely read the code you contribute to Package Y

I read this as follows: user A, who released Package X under GPL, issues a PR to Package Y which is released under MIT, and user B, the maintainer of Package Y, can not read it. Am I wrong?

(I also don't happen to believe that reading GPL code and writing unrelated MIT code is problematic. Copy-pasting GPL code in MIT code is what causes the "contagion".)

gabrielgellner commented:
@carlobaldassi I see your point. Would you object to the change

> ... they might not feel safe reading the code ...

It is a bit FUD-y, but I do think @johnmyleswhite raises an interesting point on how contagion might be perceived in projects that use GITHUB as a package manager! At least in this wording we are being less definative and just expressing potential interpreations in the community (by at least one person that has written a LOT of MIT libraries for Julia!) 

My, not a lawyer, interpretation is as follows:
If I require a MIT package in my GPL package (not LGPL) and then I sent pull requests to the MIT package as a result of needs I discover in my GPL package, I can see how this might be seen as being a derivative work (versus a developer that might contribute independently to GPL and MIT packages were there is no clear source code connection between the packages.)

Keno commented:
> If I require a MIT package in my GPL package (not LGPL) and then I sent pull requests to the MIT package as a result of needs I discover in my GPL package, I can see how this might be seen as being a derivative work (versus a developer that might contribute independently to GPL and MIT packages were there is no clear source code connection between the packages.)

This isn't the problem, as the author you are free to license your work under whatever license you wish. The problem is this: Say Gadfly were GPL, but uses Colors which is MIT. Now Gadfly has some code (say for color swatches) that is generically useful and we would like to move to Colors. To do, so, we'd need the permission of every copyright holder to Gadfly (since the code may have been written using other parts of Gadfly, etc.). Since the majority of Julia packages and julia itself are MIT licensed, this would inhibit cross-package refactoring. Realistically, I think it is only feasibly to have a GPL package if you also require copyright assignment (or at least a permissive CLA) by all contributors, so at least the repository maintainer can authorize such code movements (as opposed to having to ask for permissions from every single contributor).

wildart commented:
I'd  prefer to use SPDX identifier for licences (http://spdx.org/licenses/). So in a future we could grab any licence from their licence repo (http://git.spdx.org/license-list.git/).

gabrielgellner commented:
@wildart That seems like a cool idea, but I worry it is mission creep for this pull request. It is not clear that people are going to merge even this subset :)

If another issue is opened to use a system like you mention, it might also be interesting to look into the GitHub version this same kind of service that is integrated into the web interface (http://choosealicense.com/)

Anyone else have feelings on the language change I suggested earlier?

gabrielgellner commented:
Is there anything needed to keep this moving? Or is this a "wontfix"?

tkelman commented:
> Anyone else have feelings on the language change I suggested earlier?

Which one, specifically?

gabrielgellner commented:
@tkelman I suggested changing the additional wording added to the `README.md` to go from (the current pull request text):

> ... For example, if a user works on Package Y, which is licensed under the MIT license that is used in many community projects, they can not safely read the code you contribute to Package X because any indication that their work is derivative could lead to litigation.

To:

> For example, if a user works on Package Y, which is licensed under the MIT license that is used in many community projects, they may not feel safe reading the code you contribute to Package X because any indication that their work is derivative could lead to litigation.

I hoped that this might soften the Note to be less absolute, but still convey the perceived problems with contributing code across license boundaries.

I never made the change, but am happy to if this is felt to be more in line with the message we want to send to users. I think this is fair, but am easy going on the exact wording.

tkelman commented:
That seems okay to me.

nalimilan commented:
@gabrielgellner Could you  rename the licenses to `GPL-2.0+` and `LGPL-2.0+`? It seems important to make the "any later version" part very explicit. The `-2.0` part allows using the same identifiers as SPDX and Choose a License, which could always be useful.

gabrielgellner commented:
Happy to make these changes.

@nalimilan Should the api for selecting the license reflect the plus as well? ie do we want
```julia
Pkg.generate("foo", "GPL2+")
```
or do we leave it with the string code "GPL2"?

nalimilan commented on src/PkgDev.jl:
```diff
@@ -63,7 +63,7 @@ publish(prbranch::AbstractString="") = Entry.publish(Pkg.Dir.getmetabranch(), pr
     generate(pkg,license)
 
-Generate a new package named `pkg` with one of these license keys: `"MIT"`, `"BSD"`, `"ASL"` or `"MPL"`. If you want to make a package with a different license, you can edit it afterwards. Generate creates a git repo at `Pkg.dir(pkg)` for the package and inside it `LICENSE.md`, `README.md`, `REQUIRE`, and the julia entrypoint `$pkg/src/$pkg.jl`. Travis, AppVeyor CI configuration files `.travis.ym ...
+Generate a new package named `pkg` with one of these license keys: `"MIT"`, `"BSD"`, `"ASL"`, `"MPL"`, `"GPL2"`, `"GPL3"`, `"LGPL2"`, `"LGPL3"`. If you want to make a package with a different license, you can edit it afterwards. Generate creates a git repo at `Pkg.dir(pkg)` for the package and inside it `LICENSE.md`, `README.md`, `REQUIRE`, and the julia entrypoint `$pkg/src/$pkg.jl`. Travis, App ...
```
You need to update the names here too.

gabrielgellner replied:
Thanks for the catch. Fixed.

I can rebase when people want me to. I just find it dark magic and avoid until the end ...

nalimilan commented:
Thanks! Looks good to me. No need to rebase, and we can squash directly from GitHub.

tkelman commented on src/PkgDev.jl:
```diff
@@ -63,7 +63,7 @@ publish(prbranch::AbstractString="") = Entry.publish(Pkg.Dir.getmetabranch(), pr
     generate(pkg,license)
 
-Generate a new package named `pkg` with one of these license keys: `"MIT"`, `"BSD"`, `"ASL"`, `"MPL"`, `"GPL2"`, `"GPL3"`, `"LGPL2"`, `"LGPL3"`. If you want to make a package with a different license, you can edit it afterwards. Generate creates a git repo at `Pkg.dir(pkg)` for the package and inside it `LICENSE.md`, `README.md`, `REQUIRE`, and the julia entrypoint `$pkg/src/$pkg.jl`. Travis, App ...
+Generate a new package named `pkg` with one of these license keys: `"MIT"`, `"BSD"`, `"ASL"`, `"MPL"`, `"GPL-2.0+"`, `"GPL-3.0+"`, `"LGPL-2.1+"`, `"LGPL-3.0+"`. If you want to make a package with a different license, you can edit it afterwards. Generate creates a git repo at `Pkg.dir(pkg)` for the package and inside it `LICENSE.md`, `README.md`, `REQUIRE`, and the julia entrypoint `$pkg/src/$pkg. ...
```
we should wrap this at some point. doesn't need to be fixed in this PR though

gabrielgellner replied:
Will only take a second if this is wanted. Would I wrap at 92? Or the more readable 80?

tkelman replied:
whichever the person who does it prefers

nalimilan replied:
CONTRIBUTING.md says 92 chars.

tkelman replied:
as an upper bound, not an exact minimum

gabrielgellner replied:
Ha. I chose to use 92 My feeling is that it is easier to use an editors "reflow" if we stick to the same line wrap as for code. Which I interpreted the text of the Julia docs that says:

> Respect the line length limit used in the surrounding code.
Docstrings are edited using the same tools as code. Therefore, the same conventions should apply. It it advised to add line breaks after 92 characters

gabrielgellner commented on src/PkgDev.jl:
```diff
@@ -30,15 +30,19 @@ dir(pkg::AbstractString, args...) = normpath(dir(pkg),args...)
+Tag `commit` as version `ver` of package `pkg` and create a version  entry in `METADATA`. If
+not provided, `commit` defaults to the current  commit of the `pkg` repo. If `ver` is one of
+the symbols `:patch`,  `:minor`, `:major` the next patch, minor or major version is used. If
+`ver` is not provided, it defaults to `:patch`.
```
Note: I reflowed all the docstrings in the package for consistency. If that is to much, I can role back and be more focused.

gabrielgellner commented on src/PkgDev.jl:
```diff
@@ -118,16 +137,14 @@ end
-tagged release, but for which a tagged release is equivalent to the
-current version. You can use `Pkg.free(PkgDev.freeable())` to
-automatically free all such packages.
+Return a list of packages which are good candidates for `Pkg.free`. These are packages for
```
I changed the wording here from "Returns" to "Return" to follow the guidelines in the main Julia docs on docstrings using the "Imperative voice"

nalimilan commented:
Let's merge?

wildart commented:
:+1:
